Restructure developer onboarding (#303)
HarperFast/documentation: docs/foundations/core-concepts.md

---
title: Core Concepts
---

# Core Concepts

Before you build your first app with Harper, it helps to understand a few key ideas. These concepts show you how Harper is structured and why it’s flexible enough to power everything from a quick proof-of-concept to a production-ready platform.

## Components

**Components** are the building blocks of Harper.  
They’re JavaScript-based modules that extend Harper’s core, and they can talk directly to Harper’s [Global APIs](../reference/globals) (databases, tables, resources).

Because components can build on top of each other, they give you composability. For example, both [Applications](../developers/applications/) and [Extensions](../reference/components/built-in-extensions) are just kinds of components:

> - **Extensions** add individual capabilities, like defining tables or serving static assets.
> - **Applications** pull multiple extensions and resources together into a complete product.

Reviewer: ```suggestion
- **Plugins** add individual capabilities, like defining tables or serving static assets.
- **Applications** pull multiple plugins and resources together into a complete product.
```

:::info
💡 **Why it matters:** Instead of wiring up a backend from scratch, you can piece together pre-built functionality and get to working endpoints fast.  
:::

## Applications (a type of Component)

An **application** is a special kind of component that pulls everything together.  

> Applications rely on extensions to do the work:

Reviewer: ```suggestion
Applications rely on plugins to do the work:
```

- Use `graphqlSchema` to define your data tables.
- Add `rest` to query that data instantly.
- Plug in `static` to serve files or front-end assets.

You can even run full frameworks like [Next.js](https://github.com/HarperDB/nextjs) or [Apollo](https://github.com/HarperDB/apollo) as Harper applications.

:::info
💡 **Why it matters:** Applications are how you ship real products on Harper. They let you stitch together resources, APIs, and UI in one place.
:::

## Plugins (formerly Extensions)

> Harper has a special kind of component that isn’t meant to run standalone, but instead adds features to applications or other components. These were originally called **extensions**, and the newer API calls them **plugins**. Both terms refer to the same concept, but “plugin” is the direction going forward.

Reviewer: ```suggestion
**Plugins** are a special kind of component that are not meant to run standalone, but instead add features to applications or other components. These were originally called **extensions** (and the [extension API](../reference/components/extensions) is still supported), but the new [plugin API](../reference/components/plugins) is simultaneously a simplification and extensibility upgrade. 
```
